server/v2/appmanager/service.go

```diff
@@ -94,30 +94,41 @@ func (a AppManager[T]) VerifyBlock(ctx context.Context, height uint64, txs []T)
 	return nil
 }
 
-func (a AppManager[T]) DeliverBlock(ctx context.Context, block appmanager.BlockRequest) (*appmanager.BlockResponse, Hash, error) {
+func (a AppManager[T]) DeliverBlock(ctx context.Context, block appmanager.BlockRequest) (*appmanager.BlockResponse, Hash, []store.ChangeSet, error) {
 	currentState, err := a.db.NewStateAt(block.Height)
 	if err != nil {
-		return nil, nil, fmt.Errorf("unable to create new state for height %d: %w", block.Height, err)
+		return nil, nil, nil, fmt.Errorf("unable to create new state for height %d: %w", block.Height, err)
 	}
 
 	blockResponse, newState, err := a.stf.DeliverBlock(ctx, block, currentState)
 	if err != nil {
-		return nil, nil, fmt.Errorf("block delivery failed: %w", err)
+		return nil, nil, nil, fmt.Errorf("block delivery failed: %w", err)
 	}
 
-	// apply new state to store
-	newStateChanges, err := newState.ChangeSets()
+	changesets, err := newState.ChangeSets()
 	if err != nil {
-		return nil, nil, fmt.Errorf("change set: %w", err)
+		return nil, nil, nil, err
 	}
 
-	stateRoot, err := a.db.CommitState(newStateChanges)
+	stateRoot, err := newState.WorkingHash()
```

Comment:
I think `DeliverBlock` should not return a `hash`, for performance reasons, we can let the consensus engine pre-emptively compute the hash if needed when the AppMAnager returns it.

Reply:
How would that work exactly? E.g. for CometBFT, we need to call `WorkingHash` before `Commit` is called, since `FinalizeBlock` expects the root app hash.

Reply:
eg: 

```go
type WorkingHashProvider interface { GetWorkingHash(cs ChangeSet) }

type cometBFT struct {
   am AppManager
   whp WorkingHashProvider
}

func (c cometBFT) FinalizeBlock(ctx, req) (resp, err) {
   amReq := intoAmReq(req)
   amResp,stateChanges, err := c.am.DeliverBlock(ctx, amReq)
   if err != nil { ... }
   
   // compute hash
  workingHash := c.whp.GetWorkingHash(stateChanges)
  return intoCometResp(workingHash, amResp)
}
```

the idea is that we do not let "consensus" specific assumptions leak into appmanager because then we lose on modularity.

Reply:
> the idea is that we do not let "consensus" specific assumptions leak into appmanager because then we lose on modularity.

100% -- proposal makes sense to me.

Reply:
makes sense, ill adjust here. thanks for the comments